### DISCUSSION TITLE
Skipping properties

### DISCUSSION THREAD

**Discussion by NightProgramming** [RebeccaStevens/deepmerge-ts#50] [Ideas]
Suppose we don't need any `timestamp` properties on the merge result. Of course we could delete all the properties from the result but we would need to deeply iterate over all objects ourselves.

Instead we could just use what deepmerge already provides (deep iteration) + the possibility to skip properties:

``` JS
mergeOthers: (values, utils, meta) => {
      const { key } = meta;

      if (key === "timestamp") {
          utils.skipProperty();
          return null; // It shouldn't matter what is returned, because the property is not going to exist anyway.
      }
    }
```

Now the merge result should have no `timestamp` property at all.

**Comment by RebeccaStevens** [comment]
This would be a nice feature but I'm not sure how best to go about implementing it.

I want to keep this library functional where possible and thus nothing should rely on or cause side effects. #51 would obviously be an exception to this as margining into an object, mutates that object (and thus isn't functional).

However, if this can't nicely be done in a functional way, that doesn't necessarily rule out this feature from being implemented.

Currently to do a merge like, you would do it something so:

```ts
mergeRecords: (values, utils, meta) => {
  const valuesWithoutTimestamp = values.map(({ timestamp, ...rest }) => rest);
  return utils.defaultMergeFunctions.mergeRecords(valuesWithoutTimestamp, utils, meta);
}
```

**Comment by NightProgramming** [comment]
@RebeccaStevens When you talk about side effects you mean that `utils.skipProperty();` affects other merge operations? If so then I probably gave a bad example. Maybe this one is better:

```JS
mergeOthers: (values, utils, meta, options) => {
    if (key === "timestamp") {
          options.skipProperty();
          // Or
          options.skipProperty = true;
     }
}
```

Things done with `options` object only affect the outcome of the current custom merge. It's for local, not global settings. If performance is an issue, maybe create one object at the very beginning (when client calls `deepmerge()`) and re-use this object for every execution of the custom merger during this whole deep merge. Of course the options object then needs to be reset after every custom merger call (e.g. setting `skipProperty` to false). This should be faster than instantiating a new options object before every call of the custom merger.

**Reply by RebeccaStevens** [reply]
No. What I mean is that the function should be ["pure"](https://en.wikipedia.org/wiki/Pure_function).

So both `utils.skipProperty()`/`options.skipProperty()` and `options.skipProperty = true` would make the function non-pure as they cause an effect outside of the function. To keep the function pure, we'd want to return a value that means skip the property.

**Comment by RebeccaStevens** [comment]
@NightProgramming Implemented in #64
